File: ai-service/app/services/llm.py
"""LLM access with mandatory failover (SR-11).

Gemini 2.5 Flash is primary; Groq (Llama) is the failover. Both expose
OpenAI-compatible chat completions so a single code path serves both.
429s and timeouts fail over instead of failing the request.
"""

from dataclasses import dataclass
from typing import Literal

import httpx
from pydantic import BaseModel

from app.core.config import get_settings

Role = Literal["system", "user", "assistant"]


class ChatMessage(BaseModel):
    role: Role
    content: str


@dataclass(frozen=True)
class _Provider:
    name: str
    url: str
    model: str
    api_key: str


class LlmUnavailableError(Exception):
    def __init__(self, attempts: list[str]) -> None:
        self.attempts = attempts
        super().__init__(f"all LLM providers failed: {'; '.join(attempts)}")


def _providers() -> list[_Provider]:
    settings = get_settings()
    chain: list[_Provider] = []
    if settings.gemini_api_key:
        chain.append(
            _Provider(
                name="gemini",
                url="https://generativelanguage.googleapis.com/v1beta/openai/chat/completions",
                model="gemini-2.5-flash",
                api_key=settings.gemini_api_key,
            )
        )
    if settings.groq_api_key:
        chain.append(
            _Provider(
                name="groq",
                url="https://api.groq.com/openai/v1/chat/completions",
                model="llama-3.3-70b-versatile",
                api_key=settings.groq_api_key,
            )
        )
    return chain


async def chat_complete(
    messages: list[ChatMessage],
    *,
    temperature: float = 0.2,
    json_mode: bool = False,
) -> tuple[str, str]:
    """Returns (assistant_text, provider_name). Raises LlmUnavailableError."""
    settings = get_settings()
    attempts: list[str] = []

    for provider in _providers():
        body: dict[str, object] = {
            "model": provider.model,
            "messages": [m.model_dump() for m in messages],
            "temperature": temperature,
        }
        if json_mode:
            body["response_format"] = {"type": "json_object"}

        try:
            async with httpx.AsyncClient(timeout=settings.llm_timeout_seconds) as client:
                res = await client.post(
                    provider.url,
                    json=body,
                    headers={"Authorization": f"Bearer {provider.api_key}"},
                )
        except httpx.TimeoutException:
            attempts.append(f"{provider.name}: timeout")
            continue
        except httpx.HTTPError as exc:
            attempts.append(f"{provider.name}: {exc.__class__.__name__}")
            continue

        if res.status_code == 429:
            attempts.append(f"{provider.name}: rate limited (429)")
            continue
        if res.status_code != 200:
            attempts.append(f"{provider.name}: HTTP {res.status_code}")
            continue

        data = res.json()
        text = (data.get("choices") or [{}])[0].get("message", {}).get("content")
        if not text:
            attempts.append(f"{provider.name}: empty response")
            continue
        return text, provider.name

    if not attempts:
        attempts.append("no providers configured (set GEMINI_API_KEY / GROQ_API_KEY)")
    raise LlmUnavailableError(attempts)

File: supabase/functions/_shared/llm.ts
// Shared LLM client for edge functions.
// SR-11 / CLAUDE.md §6: every LLM call is wrapped with primary→secondary
// failover (Gemini → Groq) and explicit 429/timeout handling. Both providers
// expose OpenAI-compatible chat completions, so one code path serves both.

export interface ChatMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

interface Provider {
  name: string;
  url: string;
  model: string;
  apiKey: string | undefined;
}

const TIMEOUT_MS = 25_000;

function providers(): Provider[] {
  return [
    {
      name: "gemini",
      url: "https://generativelanguage.googleapis.com/v1beta/openai/chat/completions",
      model: "gemini-2.5-flash",
      apiKey: Deno.env.get("GEMINI_API_KEY"),
    },
    {
      name: "groq",
      url: "https://api.groq.com/openai/v1/chat/completions",
      model: "llama-3.3-70b-versatile",
      apiKey: Deno.env.get("GROQ_API_KEY"),
    },
  ];
}

export class LlmUnavailableError extends Error {
  constructor(public attempts: string[]) {
    super(`All LLM providers failed: ${attempts.join("; ")}`);
  }
}

/**
 * Chat completion with failover. Returns the assistant text.
 * `jsonMode` asks the provider for a JSON object response.
 */
export async function chatComplete(
  messages: ChatMessage[],
  opts: { jsonMode?: boolean; temperature?: number } = {},
): Promise<{ text: string; provider: string }> {
  const attempts: string[] = [];

  for (const p of providers()) {
    if (!p.apiKey) {
      attempts.push(`${p.name}: no API key configured`);
      continue;
    }
    try {
      const res = await fetch(p.url, {
        method: "POST",
        signal: AbortSignal.timeout(TIMEOUT_MS),
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${p.apiKey}`,
        },
        body: JSON.stringify({
          model: p.model,
          messages,
          temperature: opts.temperature ?? 0.2,
          ...(opts.jsonMode ? { response_format: { type: "json_object" } } : {}),
        }),
      });

      if (res.status === 429) {
        attempts.push(`${p.name}: rate limited (429)`);
        continue; // fail over
      }
      if (!res.ok) {
        attempts.push(`${p.name}: HTTP ${res.status}`);
        continue;
      }

      const data = await res.json();
      const text: string | undefined = data?.choices?.[0]?.message?.content;
      if (!text) {
        attempts.push(`${p.name}: empty response`);
        continue;
      }
      return { text, provider: p.name };
    } catch (err) {
      const kind = err instanceof DOMException && err.name === "TimeoutError"
        ? "timeout"
        : String(err);
      attempts.push(`${p.name}: ${kind}`);
    }
  }

  throw new LlmUnavailableError(attempts);
}

/**
 * Text embedding via Gemini (text-embedding-004, 768 dims — matches the
 * vector(768) columns). No Groq fallback: Groq has no embeddings endpoint,
 * so callers must handle LlmUnavailableError by retrying later.
 */
export async function embedText(text: string): Promise<number[]> {
  const apiKey = Deno.env.get("GEMINI_API_KEY");
  if (!apiKey) throw new LlmUnavailableError(["gemini: no API key configured"]);

  const res = await fetch(
    "https://generativelanguage.googleapis.com/v1beta/openai/embeddings",
    {
      method: "POST",
      signal: AbortSignal.timeout(TIMEOUT_MS),
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({ model: "text-embedding-004", input: text }),
    },
  );

  if (!res.ok) {
    throw new LlmUnavailableError([`gemini embeddings: HTTP ${res.status}`]);
  }
  const data = await res.json();
  const vec: number[] | undefined = data?.data?.[0]?.embedding;
  if (!vec) throw new LlmUnavailableError(["gemini embeddings: empty response"]);
  return vec;
}

export const corsHeaders: Record<string, string> = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Headers":
    "authorization, x-client-info, apikey, content-type",
};

export function jsonResponse(body: unknown, status = 200): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { ...corsHeaders, "Content-Type": "application/json" },
  });
}
